geomesa-core/src/main/scala/org/locationtech/geomesa/core/csv/package.scala

```diff
@@ -39,84 +39,80 @@ import scala.collection.JavaConversions._
-        case Some(x) => x._2
-        case None    => 's'   // should get to this anyway as StringParser is guaranteed to succeed
-      }
+  def csvToFeatures(csvFile: File,
+                    hasHeader: Boolean,
+                    typeSchema: TypeSchema): SimpleFeatureCollection = {
+    val sft = SimpleFeatureTypes.createType(typeSchema.name, typeSchema.schema)
+    buildFeatureCollection(csvFile, hasHeader, sft, typeSchema.latLonFields)
+  }
 
+  protected[csv] def tryParsers(rawData: TraversableOnce[String]): Seq[CSVParser[_]] = {
+    def tryAllParsers(datum: String) =
+      CSVParser.parsers.find(_.parse(datum).isSuccess).getOrElse(StringParser)
     rawData.map(tryAllParsers).toSeq
   }
-  
-  def sampleRecords(records: jIterator[CSVRecord], hasHeader: Boolean): (Seq[String], CSVRecord) =
-    if (hasHeader) {
-      val header = records.next
-      val record = records.next
-      (header.toSeq, record)
-    } else {
-      val record = records.next
-      val header = Seq.tabulate(record.size()) { n => s"C$n" }
-      (header, record)
-    }
 
-  def guessTypes(name: String,
+  protected[csv] def guessHeaders(record: CSVRecord, hasHeader: Boolean = true): Seq[String] =
+    if (hasHeader) record.toSeq else Seq.tabulate(record.size())(n => s"C$n")
+
+  protected[csv] def guessTypes(name: String,
                  csvReader: Reader,
                  hasHeader: Boolean = true,
-                 format: CSVFormat = CSVFormat.DEFAULT): TypeSchema = {
-    val records = format.parse(csvReader).iterator
-    val (header, record) = sampleRecords(records, hasHeader)
-    val typeChars = typeData(record.iterator)
+                 format: CSVFormat = CSVFormat.DEFAULT,
+                 numSamples: Int = 5): TypeSchema = {
+    assert(numSamples > 0)
+    val records = format.parse(csvReader).iterator.take(numSamples + 1).toSeq
+    assert(records.size > 1 || (!hasHeader && records.size > 0))
+
+    val headers = guessHeaders(records(0), hasHeader)
+    val sample = records.drop(1)
+
+    // make sure type chars are valid for first few records
+    val parsers = sample.map(tryParsers(_)).reduceLeft { (pc1, pc2)  =>
+      pc1.zip(pc2).map { case (p1, p2) => if (p1 == p2) p1 else StringParser }
+    }
 
     val sftb = new SftBuilder
-    var defaultDateSet = false
     var defaultGeomSet = false
-    for ((field, c) <- header.zip(typeChars)) { c match {
-      case 'i' =>
-        sftb.intType(field)
-      case 'd' =>
-        sftb.doubleType(field)
-      case 't' =>
-        sftb.date(field)
-        if (!defaultDateSet) {
-          sftb.withDefaultDtg(field)
-          defaultDateSet = true
-        }
-      case 'p' =>
-        if (defaultGeomSet) sftb.geometry(field)
-        else {
-          sftb.point(field, default = true)
-          defaultGeomSet = true
-        }
-      case 's' =>
-        sftb.stringType(field)
-    }}
+    headers.zip(parsers).foreach { case (field, parser) =>
```

Comment:
The goal was to try to not duplicate/split logic about types. Since each parser already knows it's type, it knows how to define it with the SftBuilder. It made more sense to me to keep all the type logic in the parsers, vs splitting some of it out to here based on the parser 'flag'. Also, once I added all the other geometry types, the flags became unwieldy.
The SftBuilder is still declared locally, it gets passed in to the parser to call the correct method.